server/plugins/secrets/encrypted.go

```diff
@@ -0,0 +1,248 @@
+	}
+	err = ess.decrypt(secret)
+	if err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func (ess *encryptedSecretService) OrgSecretList(owner int64, listOpt *model.ListOptions) ([]*model.Secret, error) {
+	var err error
+	secrets, err := ess.secretSvc.OrgSecretList(owner, listOpt)
+	if err != nil {
+		return nil, err
+	}
+	err = ess.decryptList(secrets)
+	if err != nil {
+		return nil, err
+	}
+	return secrets, nil
+}
+
+func (ess *encryptedSecretService) OrgSecretCreate(owner int64, in *model.Secret) error {
+	var err error
+	err = ess.encrypt(in)
+	if err != nil {
+		return err
+	}
+	err = ess.secretSvc.OrgSecretCreate(owner, in)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ess *encryptedSecretService) OrgSecretUpdate(owner int64, in *model.Secret) error {
+	var err error
+	err = ess.encrypt(in)
+	if err != nil {
+		return err
+	}
+	err = ess.secretSvc.OrgSecretUpdate(owner, in)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ess *encryptedSecretService) OrgSecretDelete(owner int64, name string) error {
+	return ess.secretSvc.OrgSecretDelete(owner, name)
+}
+
+func (ess *encryptedSecretService) GlobalSecretFind(owner string) (*model.Secret, error) {
+	var err error
+	secret, err := ess.secretSvc.GlobalSecretFind(owner)
+	if err != nil {
+		return nil, err
+	}
+	err = ess.decrypt(secret)
+	if err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func (ess *encryptedSecretService) GlobalSecretList(listOpt *model.ListOptions) ([]*model.Secret, error) {
+	var err error
+	secrets, err := ess.secretSvc.GlobalSecretList(listOpt)
+	if err != nil {
+		return nil, err
+	}
+	err = ess.decryptList(secrets)
+	if err != nil {
+		return nil, err
+	}
+	return secrets, nil
+}
+
+func (ess *encryptedSecretService) GlobalSecretCreate(in *model.Secret) error {
+	var err error
+	err = ess.encrypt(in)
+	if err != nil {
+		return err
+	}
+	err = ess.secretSvc.GlobalSecretCreate(in)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ess *encryptedSecretService) GlobalSecretUpdate(in *model.Secret) error {
+	var err error
+	err = ess.encrypt(in)
+	if err != nil {
+		return err
+	}
+	err = ess.secretSvc.GlobalSecretUpdate(in)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ess *encryptedSecretService) GlobalSecretDelete(name string) error {
+	return ess.secretSvc.GlobalSecretDelete(name)
+}
+
+func (ess *encryptedSecretService) encrypt(secret *model.Secret) error {
+	log.Debug().Int64("id", secret.ID).Str("name", secret.Name).Msg("encryption")
+
+	encryptedValue, err := ess.encryptionSvc.Encrypt(secret.Value, secret.Name)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt secret id=%d: %w", secret.ID, err)
+	}
+	secret.Value = encryptedValue
+	return nil
+}
+
+func (ess *encryptedSecretService) decrypt(secret *model.Secret) error {
+	log.Debug().Int64("id", secret.ID).Str("name", secret.Name).Msg("decryption")
+
+	decryptedValue, err := ess.encryptionSvc.Decrypt(secret.Value, secret.Name)
```

Comment:
Could we add some logic here that detects the encryption method. So it would know if a secret was encrypted using aes or maybe tink, ... And if the secret wasn't encrypted at all (before enabling the encryption) it will simply skip the decryption

Reply:
> Could we add some logic here that detects the encryption method. So it would know if a secret was encrypted using aes or maybe tink

It's overcomplicated, I think. Admin should know, what encryption he enables and set up WP accordingly. So, we know encryption method at runtime.

> And if the secret wasn't encrypted at all (before enabling the encryption) it will simply skip the decryption

Then what's the point of encrypted decorator? In that case plain `SecretService` should be used.

I would just get back migration logic. From plain to encrypted according to settings (AES, Tink). I'll do it in this PR.

As for more complicated migrations like AES -> Tink, Encrypted -> Plain, I would use CLI or maybe some button in admin panel. But this is out of scope of this PR.

Reply:
I had sth simple like this common-helper in mind which would just support decrypting plain secrets as well. It would also allow us to not do an active migration and instead do a migration on demand each time a user adds a new secrets or updates an existing one.

```go
func decryptValue(secret *Secret) (string, string) {
	if strings.HasPrefix(secret.Value, "_aes_") {
		return "aes", strings.TrimPrefix(secret.Value, "_aes_")
	}

	return "plain", secret.Value
}

func decrypt(secret *Secret) error {
	method, value := decryptValue(secret)
	if method == "plain" {
		// do nothing
	} else if method == "aes" {
		// do aes stuff
		value, err := aesMagic(value)
		secret.Value = value
		return err
	}

	return fmt.Errorf("unknown secret encryption method")
}
```

Reply:
I had sth simple like this common-helper in mind which would just support decrypting plain secrets as well. It would also allow us to not do an active migration and instead do a migration on demand each time a user adds a new secrets or updates an existing one.

```go
func decryptValue(secret *Secret) (string, string) {
	if strings.HasPrefix(secret.Value, "_aes_") {
		return "aes", strings.TrimPrefix(secret.Value, "_aes_")
	}

	return "plain", secret.Value
}

func decrypt(secret *Secret) error {
	method, value := decryptValue(secret)
	if method == "plain" {
		// do nothing
	} else if method == "aes" {
		// do aes stuff
		value, err := aesMagic(value)
		secret.Value = value
		return err
	}

	return fmt.Errorf("unknown secret encryption method")
}
```

Reply:
I've added encoding/decoding as you sad. At least it looks more readable in DB.

> not do an active migration and instead do a migration on demand each time a user adds a new secrets or updates an existing one

However, as an administrator I prefer active migration. I schedule time to evaluate new feature and then to roll it out. And I would do it carefully under my control. I would not want to have issues month later after enabling the feature.